File: ffi/secretary-ffi-uniffi/tests/swift/SmokeShareBlock.swift
// B.4d share_block assertions for the Swift smoke runner.
//
// share_block extends a block's recipient list. v1 single-author: only the
// vault owner can share blocks they authored. The assertions below use
// golden_vault_001 as the owner and golden_vault_002's owner card as
// "Alice" — distinct identities pre-built by the same fixture builder.
//
// NotAuthor is NOT asserted at this layer: reaching it requires staging
// cross-vault manifest content (one vault's manifest must list a block
// authored elsewhere), which open_vault_with_password rejects via
// vault.toml ↔ manifest consistency. Pinned at the bridge unit-test
// layer (`vault_error_not_author_from_core_preserves_fingerprints_as_hex`)
// and the core integration layer
// (`core/tests/share_block.rs::share_block_non_author_rejected`); will
// be exercised end-to-end by Sub-project C's sync layer.

import Foundation

func runShareBlockAsserts(env: SmokeEnv) {
    // Assert 27: share_block happy path — owner saves a block, owner shares
    // with Alice, manifest entry now lists 2 recipients.
    do {
        let aliceBytes = try _aliceCardBytes(env: env)
        let (identity, manifest, tmp) = try _freshWritableVault(env: env)
        defer { identity.wipe() }
        defer { manifest.wipe() }
        defer { try? FileManager.default.removeItem(at: tmp) }

        try saveBlock(
            identity: identity,
            manifest: manifest,
            input: BlockInput(
                blockUuid: shareBlockBlockUuid,
                blockName: "shared",
                records: [
                    RecordInput(
                        recordUuid: shareBlockRecordUuid,
                        recordType: "",
                        tags: [],
                        fields: [FieldInput(name: "k", value: .text(text: "v"))]
                    ),
                ]
            ),
            deviceUuid: shareBlockDeviceUuid,
            nowMs: 1_000
        )
        guard let ownerBytes = try manifest.ownerCardBytes() else {
            check(false, "owner card bytes nil before share")
            exit(1)
        }
        try shareBlock(
            identity: identity,
            manifest: manifest,
            blockUuid: shareBlockBlockUuid,
            existingRecipientCards: [ownerBytes],
            newRecipient: aliceBytes,
            deviceUuid: shareBlockDeviceUuid,
            nowMs: 2_000
        )
        let summary = manifest.findBlock(blockUuid: shareBlockBlockUuid)
        check(
            summary?.recipientUuids.count == 2,
            "share_block insert → manifest grows to 2 recipients (got \(summary?.recipientUuids.count ?? -1))"
        )
    } catch {
        check(false, "share_block happy path threw \(error), expected to succeed")
    }

    // Assert 28: share_block to the same recipient twice → RecipientAlreadyPresent.
    do {
        let aliceBytes = try _aliceCardBytes(env: env)
        let (identity, manifest, tmp) = try _freshWritableVault(env: env)
        defer { identity.wipe() }
        defer { manifest.wipe() }
        defer { try? FileManager.default.removeItem(at: tmp) }

        try saveBlock(
            identity: identity,
            manifest: manifest,
            input: BlockInput(
                blockUuid: shareBlockBlockUuid, blockName: "x", records: []
            ),
            deviceUuid: shareBlockDeviceUuid,
            nowMs: 1_000
        )
        let ownerBytes = try manifest.ownerCardBytes()!
        try shareBlock(
            identity: identity,
            manifest: manifest,
            blockUuid: shareBlockBlockUuid,
            existingRecipientCards: [ownerBytes],
            newRecipient: aliceBytes,
            deviceUuid: shareBlockDeviceUuid,
            nowMs: 2_000
        )
        do {
            try shareBlock(
                identity: identity,
                manifest: manifest,
                blockUuid: shareBlockBlockUuid,
                existingRecipientCards: [ownerBytes, aliceBytes],
                newRecipient: aliceBytes,
                deviceUuid: shareBlockDeviceUuid,
                nowMs: 3_000
            )
            check(false, "duplicate share_block should have thrown VaultError.RecipientAlreadyPresent")
        } catch let e as VaultError {
            if case .RecipientAlreadyPresent = e {
                check(true, "share_block duplicate alice → RecipientAlreadyPresent")
            } else {
                check(false, "duplicate share_block threw wrong VaultError variant: \(e)")
            }
        }
    } catch {
        check(false, "share_block duplicate-recipient setup threw \(error)")
    }

    // Assert 29: share_block with empty existing_recipient_cards while the
    // block has the owner as a recipient → MissingRecipientCard.
    do {
        let aliceBytes = try _aliceCardBytes(env: env)
        let (identity, manifest, tmp) = try _freshWritableVault(env: env)
        defer { identity.wipe() }
        defer { manifest.wipe() }
        defer { try? FileManager.default.removeItem(at: tmp) }

        try saveBlock(
            identity: identity,
            manifest: manifest,
            input: BlockInput(
                blockUuid: shareBlockBlockUuid, blockName: "x", records: []
            ),
            deviceUuid: shareBlockDeviceUuid,
            nowMs: 1_000
        )
        do {
            try shareBlock(
                identity: identity,
                manifest: manifest,
                blockUuid: shareBlockBlockUuid,
                existingRecipientCards: [],
                newRecipient: aliceBytes,
                deviceUuid: shareBlockDeviceUuid,
                nowMs: 2_000
            )
            check(false, "share with empty existing list should have thrown MissingRecipientCard")
        } catch let e as VaultError {
            if case let .MissingRecipientCard(fp) = e {
                check(
                    fp.count == 32,
                    "share_block missing card → MissingRecipientCard(\(fp))"
                )
            } else {
                check(false, "missing-existing-card threw wrong VaultError variant: \(e)")
            }
        }
    } catch {
        check(false, "share_block missing-existing-card setup threw \(error)")
    }

    // Assert 30: share_block with garbage card bytes → CardDecodeFailure.
    do {
        let aliceBytes = try _aliceCardBytes(env: env)
        let (identity, manifest, tmp) = try _freshWritableVault(env: env)
        defer { identity.wipe() }
        defer { manifest.wipe() }
        defer { try? FileManager.default.removeItem(at: tmp) }

        try saveBlock(
            identity: identity,
            manifest: manifest,
            input: BlockInput(
                blockUuid: shareBlockBlockUuid, blockName: "x", records: []
            ),
            deviceUuid: shareBlockDeviceUuid,
            nowMs: 1_000
        )
        let garbage = Data(repeating: 0xff, count: 8)
        do {
            try shareBlock(
                identity: identity,
                manifest: manifest,
                blockUuid: shareBlockBlockUuid,
                existingRecipientCards: [garbage],
                newRecipient: aliceBytes,
                deviceUuid: shareBlockDeviceUuid,
                nowMs: 2_000
            )
            check(false, "garbage existing card should have thrown CardDecodeFailure")
        } catch let e as VaultError {
            if case .CardDecodeFailure = e {
                check(true, "share_block garbage existing → CardDecodeFailure")
            } else {
                check(false, "garbage existing threw wrong VaultError variant: \(e)")
            }
        }
    } catch {
        check(false, "share_block card-decode-failure setup threw \(error)")
    }
}

File: ffi/secretary-ffi-uniffi/tests/kotlin/SmokeShareBlock.kt
// B.4d share_block assertions for the Kotlin smoke runner.
//
// share_block extends a block's recipient list. Same NotAuthor exclusion
// rationale as the Swift smoke runner — cross-vault manifest staging is
// impractical at this layer (pinned at bridge unit-test layer instead).

import uniffi.secretary.BlockInput
import uniffi.secretary.FieldInput
import uniffi.secretary.FieldInputValue
import uniffi.secretary.RecordInput
import uniffi.secretary.VaultException
import uniffi.secretary.importContactCard
import uniffi.secretary.saveBlock
import uniffi.secretary.shareBlock
import uniffi.secretary.shareBlockTo

fun runShareBlockAsserts(env: SmokeEnv) {
    // Assert 28: share_block happy path — owner saves, owner shares with
    // alice, manifest entry grows from 1 to 2 recipients.
    var shareTmp: java.nio.file.Path? = null
    try {
        val aliceBytes = aliceCardBytes(env)
        val (out, tmp) = freshWritableVault(env)
        shareTmp = tmp
        out.identity.use { id ->
            out.manifest.use { mf ->
                saveBlock(
                    id, mf,
                    BlockInput(
                        blockUuid = SHARE_BLOCK_BLOCK_UUID,
                        blockName = "shared",
                        records = listOf(
                            RecordInput(
                                SHARE_BLOCK_RECORD_UUID,
                                "",
                                emptyList(),
                                listOf(FieldInput("k", FieldInputValue.Text("v"))),
                            ),
                        ),
                    ),
                    SHARE_BLOCK_DEVICE_UUID, 1_000UL,
                )
                val ownerBytes = mf.ownerCardBytes()!!
                shareBlock(
                    id, mf, SHARE_BLOCK_BLOCK_UUID,
                    listOf(ownerBytes), aliceBytes,
                    SHARE_BLOCK_DEVICE_UUID, 2_000UL,
                )
                val summary = mf.findBlock(SHARE_BLOCK_BLOCK_UUID)
                check(
                    summary?.recipientUuids?.size == 2,
                    "share_block insert → manifest grows to 2 recipients (got ${summary?.recipientUuids?.size})",
                )
            }
        }
    } catch (e: Throwable) {
        check(false, "share_block happy path threw $e, expected to succeed")
    } finally {
        shareTmp?.let { cleanupTempVault(it) }
    }

    // Assert 29: duplicate share → RecipientAlreadyPresent.
    shareTmp = null
    try {
        val aliceBytes = aliceCardBytes(env)
        val (out, tmp) = freshWritableVault(env)
        shareTmp = tmp
        out.identity.use { id ->
            out.manifest.use { mf ->
                saveBlock(
                    id, mf,
                    BlockInput(SHARE_BLOCK_BLOCK_UUID, "x", emptyList()),
                    SHARE_BLOCK_DEVICE_UUID, 1_000UL,
                )
                val ownerBytes = mf.ownerCardBytes()!!
                shareBlock(
                    id, mf, SHARE_BLOCK_BLOCK_UUID,
                    listOf(ownerBytes), aliceBytes,
                    SHARE_BLOCK_DEVICE_UUID, 2_000UL,
                )
                try {
                    shareBlock(
                        id, mf, SHARE_BLOCK_BLOCK_UUID,
                        listOf(ownerBytes, aliceBytes), aliceBytes,
                        SHARE_BLOCK_DEVICE_UUID, 3_000UL,
                    )
                    check(false, "duplicate share should have thrown RecipientAlreadyPresent")
                } catch (e: VaultException.RecipientAlreadyPresent) {
                    check(true, "share_block duplicate alice → RecipientAlreadyPresent")
                }
            }
        }
    } catch (e: Throwable) {
        // Re-raise unless already handled by the inner check().
        if (e !is VaultException.RecipientAlreadyPresent) {
            check(false, "share_block duplicate-recipient setup threw $e")
        }
    } finally {
        shareTmp?.let { cleanupTempVault(it) }
    }

    // Assert 30: empty existing list → MissingRecipientCard.
    shareTmp = null
    try {
        val aliceBytes = aliceCardBytes(env)
        val (out, tmp) = freshWritableVault(env)
        shareTmp = tmp
        out.identity.use { id ->
            out.manifest.use { mf ->
                saveBlock(
                    id, mf,
                    BlockInput(SHARE_BLOCK_BLOCK_UUID, "x", emptyList()),
                    SHARE_BLOCK_DEVICE_UUID, 1_000UL,
                )
                try {
                    shareBlock(
                        id, mf, SHARE_BLOCK_BLOCK_UUID,
                        emptyList(), aliceBytes,
                        SHARE_BLOCK_DEVICE_UUID, 2_000UL,
                    )
                    check(false, "empty existing list should have thrown MissingRecipientCard")
                } catch (e: VaultException.MissingRecipientCard) {
                    check(
                        e.recipientFingerprintHex.length == 32,
                        "share_block missing card → MissingRecipientCard(${e.recipientFingerprintHex})",
                    )
                }
            }
        }
    } catch (e: Throwable) {
        if (e !is VaultException.MissingRecipientCard) {
            check(false, "share_block missing-existing-card setup threw $e")
        }
    } finally {
        shareTmp?.let { cleanupTempVault(it) }
    }

    // Assert 32 (#206): verified safe path — import Alice by card bytes, then
    // share_block_to by UUID; manifest grows to 2 recipients. A second import
    // of the same card → ContactAlreadyExists.
    shareTmp = null
    try {
        val aliceBytes = aliceCardBytes(env)
        val (out, tmp) = freshWritableVault(env)
        shareTmp = tmp
        out.identity.use { id ->
            out.manifest.use { mf ->
                saveBlock(
                    id, mf,
                    BlockInput(SHARE_BLOCK_BLOCK_UUID, "shared", emptyList()),
                    SHARE_BLOCK_DEVICE_UUID, 1_000UL,
                )
                val summary = importContactCard(mf, aliceBytes)
                check(summary.contactUuid.size == 16, "import_contact_card -> 16-byte uuid")
                shareBlockTo(id, mf, SHARE_BLOCK_BLOCK_UUID,
                    summary.contactUuid, SHARE_BLOCK_DEVICE_UUID, 2_000UL)
                val entry = mf.findBlock(SHARE_BLOCK_BLOCK_UUID)
                check(
                    entry?.recipientUuids?.size == 2,
                    "share_block_to -> 2 recipients (got ${entry?.recipientUuids?.size})",
                )
                try {
                    importContactCard(mf, aliceBytes)
                    check(false, "duplicate import should throw ContactAlreadyExists")
                } catch (e: VaultException.ContactAlreadyExists) {
                    check(true, "duplicate import -> ContactAlreadyExists")
                }
            }
        }
    } catch (e: Throwable) {
        // Unconditional, mirroring the Swift smoke: ContactAlreadyExists is the
        // very variant the #206 guard produces, so a stray throw of it (e.g.
        // shareBlockTo erroneously rejecting) must fail the test, not be
        // swallowed. The expected duplicate-import throw is caught inline above.
        check(false, "#206 safe-path smoke threw $e")
    } finally {
        shareTmp?.let { cleanupTempVault(it) }
    }

    // Assert 31: garbage card bytes → CardDecodeFailure.
    shareTmp = null
    try {
        val aliceBytes = aliceCardBytes(env)
        val (out, tmp) = freshWritableVault(env)
        shareTmp = tmp
        out.identity.use { id ->
            out.manifest.use { mf ->
                saveBlock(
                    id, mf,
                    BlockInput(SHARE_BLOCK_BLOCK_UUID, "x", emptyList()),
                    SHARE_BLOCK_DEVICE_UUID, 1_000UL,
                )
                val garbage = ByteArray(8) { 0xff.toByte() }
                try {
                    shareBlock(
                        id, mf, SHARE_BLOCK_BLOCK_UUID,
                        listOf(garbage), aliceBytes,
                        SHARE_BLOCK_DEVICE_UUID, 2_000UL,
                    )
                    check(false, "garbage existing should have thrown CardDecodeFailure")
                } catch (e: VaultException.CardDecodeFailure) {
                    check(true, "share_block garbage existing → CardDecodeFailure")
                }
            }
        }
    } catch (e: Throwable) {
        if (e !is VaultException.CardDecodeFailure) {
            check(false, "share_block card-decode-failure setup threw $e")
        }
    } finally {
        shareTmp?.let { cleanupTempVault(it) }
    }
}
